props/saber_fett263_buttons.h

```diff
@@ -293,248 +664,3398 @@ class SaberFett263Buttons : public virtual PropBase {
+      case MENU_BLADE_STYLE:
+        sound_queue_.Play(SoundToPlay("mblade.wav"));
+        break;
+      case MENU_BLADE_LENGTH:
+        sound_queue_.Play(SoundToPlay("mblade.wav"));
+        break;
+#endif
+      case MENU_GESTURE_SUB:
+        sound_queue_.Play(SoundToPlay("mgestsub.wav"));
+        break;
+      case MENU_SETTING_SUB:
+        sound_queue_.Play(SoundToPlay("msetsub.wav"));
+        break;
+      case MENU_EFFECT:
+        sound_queue_.Play(SoundToPlay("meffect.wav"));
+        break;
+      case MENU_STYLE_SETTING_SUB:
+        sound_queue_.Play(SoundToPlay("mstylstm.wav"));
+        break;
+    }    
+  }
+
+  enum EditColorMode {
+    NONE,
+    CC_COLOR_LIST,
+#ifdef FETT263_EDIT_MODE_MENU
+    COLOR_LIST,
+    EDIT_BLACK,
+    EDIT_COLOR,
+    EDIT_WHITE,
+    ZOOM_COLOR,
+#endif
+  };
+
+#define CC_NEW_COLOR(N) NewColor(N,1);
+
+// Update Style
+  virtual void UpdateStyle(int preset_num) {
+    TRACE(PROP, "start");
+    SaveColorChangeIfNeeded();
+    // First free all styles, then allocate new ones to avoid memory
+    // fragmentation.
+#define UNSET_BLADE_STYLE(N) \
+    delete current_config->blade##N->UnSetStyle();
+    ONCEPERBLADE(UNSET_BLADE_STYLE)
+    current_preset_.SetPreset(preset_num);
+#ifdef DYNAMIC_BLADE_LENGTH
+    savestate_.ReadINIFromSaveDir("curstate");
+#define WRAP_BLADE_SHORTERNER(N) \
+    if (savestate_.blade##N##len != -1 && savestate_.blade##N##len != current_config->blade##N->num_leds()) { \
+      tmp = new BladeShortenerWrapper(savestate_.blade##N##len, tmp);	\
+    }
+#else
+#define WRAP_BLADE_SHORTERNER(N)
+#endif
+
+    
+#define SET_BLADE_STYLE(N) do {						\
+    BladeStyle* tmp = style_parser.Parse(current_preset_.current_style##N.get()); \
+    WRAP_BLADE_SHORTERNER(N)                                            \
+    current_config->blade##N->SetStyle(tmp);				\
+  } while (0);
+
+    ONCEPERBLADE(SET_BLADE_STYLE)
+
+#ifdef SAVE_COLOR_CHANGE
+    SaberBase::SetVariation(current_preset_.variation);
+#else
+    SaberBase::SetVariation(0);
+#endif
+    TRACE(PROP, "end");
+  }
+
+// Update Font / Save Style in Edit Mode, skips Preon effect (except for Preon Editing previews) using FastOn
+  virtual void UpdateFont(int preset_num) {
+    TRACE(PROP, "start");
+    bool on = SaberBase::IsOn();
+    if (on) Off();
+    SaveColorChangeIfNeeded();
+    // First free all styles, then allocate new ones to avoid memory
+    // fragmentation.
+#define UNSET_BLADE_STYLE(N) \
+    delete current_config->blade##N->UnSetStyle();
+    ONCEPERBLADE(UNSET_BLADE_STYLE)
+    current_preset_.SetPreset(preset_num);
+#define SET_BLADE_STYLE(N) \
+    current_config->blade##N->SetStyle(style_parser.Parse(current_preset_.current_style##N.get()));
+    ONCEPERBLADE(SET_BLADE_STYLE)
+    chdir(current_preset_.font.get());
+#ifdef SAVE_COLOR_CHANGE
+    SaberBase::SetVariation(current_preset_.variation);
+#else
+    SaberBase::SetVariation(0);
+#endif
+    if (on) {
+#ifdef FETT263_EDIT_MODE_MENU
```

Comment:
This isn't really the right place to check if we should play preon or not.
I suggest taking a "bool faston = true" argument to this function and then figure it out where we call this function instead.

(This is called a layering violation, when a utility function depends on code that belongs to a layer higher up in the code hierarchy.)

Reply:
fixed